packages/@react-spectrum/s2/src/CardView.tsx

```diff
@@ -0,0 +1,513 @@
+        bestDistance = distance;
+        bestKey = candidate.key;
+      }
+    }
+
+    return bestKey;
+  }
+}
+
+const layoutOptions = {
+  XS: {
+    compact: {
+      minSpace: new Size(6, 6),
+      minItemSize: new Size(100, 100),
+      maxItemSize: new Size(140, 140)
+    },
+    regular: {
+      minSpace: new Size(8, 8),
+      minItemSize: new Size(100, 100),
+      maxItemSize: new Size(140, 140)
+    },
+    spacious: {
+      minSpace: new Size(12, 12),
+      minItemSize: new Size(100, 100),
+      maxItemSize: new Size(140, 140)
+    }
+  },
+  S: {
+    compact: {
+      minSpace: new Size(8, 8),
+      minItemSize: new Size(150, 150),
+      maxItemSize: new Size(210, 210)
+    },
+    regular: {
+      minSpace: new Size(12, 12),
+      minItemSize: new Size(150, 150),
+      maxItemSize: new Size(210, 210)
+    },
+    spacious: {
+      minSpace: new Size(16, 16),
+      minItemSize: new Size(150, 150),
+      maxItemSize: new Size(210, 210)
+    }
+  },
+  M: {
+    compact: {
+      minSpace: new Size(12, 12),
+      minItemSize: new Size(200, 200),
+      maxItemSize: new Size(280, 280)
+    },
+    regular: {
+      minSpace: new Size(16, 16),
+      minItemSize: new Size(200, 200),
+      maxItemSize: new Size(280, 280)
+    },
+    spacious: {
+      minSpace: new Size(20, 20),
+      minItemSize: new Size(200, 200),
+      maxItemSize: new Size(280, 280)
+    }
+  },
+  L: {
+    compact: {
+      minSpace: new Size(16, 16),
+      minItemSize: new Size(270, 270),
+      maxItemSize: new Size(370, 370)
+    },
+    regular: {
+      minSpace: new Size(20, 20),
+      minItemSize: new Size(270, 270),
+      maxItemSize: new Size(370, 370)
+    },
+    spacious: {
+      minSpace: new Size(24, 24),
+      minItemSize: new Size(270, 270),
+      maxItemSize: new Size(370, 370)
+    }
+  },
+  XL: {
+    compact: {
+      minSpace: new Size(20, 20),
+      minItemSize: new Size(340, 340),
+      maxItemSize: new Size(460, 460)
+    },
+    regular: {
+      minSpace: new Size(24, 24),
+      minItemSize: new Size(340, 340),
+      maxItemSize: new Size(460, 460)
+    },
+    spacious: {
+      minSpace: new Size(28, 28),
+      minItemSize: new Size(340, 340),
+      maxItemSize: new Size(460, 460)
+    }
+  }
+};
+
+const cardViewStyles = style({
+  overflowY: {
+    default: 'auto',
+    isLoading: 'hidden'
+  },
+  scrollPadding: '[18px]',
+  display: {
+    isEmpty: 'flex'
+  },
+  flexDirection: 'column',
+  alignItems: 'center',
+  justifyContent: 'center',
+  outlineStyle: 'none'
+}, getAllowedOverrides({height: true}));
+
+export function CardView<T extends object>(props: CardViewProps<T>) {
+  let {children, layout: layoutName = 'grid', size = 'M', density = 'regular', variant = 'primary', selectionStyle = 'checkbox', UNSAFE_className = '', UNSAFE_style, styles, ...otherProps} = props;
+  let options = layoutOptions[size][density];
+  let layout = useMemo(() => {
+    variant; // needed to invalidate useMemo
```

Comment:
i think this is preferable over eslint-ignoring the entire dep array, should update this in other places. outside scope of pr